### DISCUSSION TITLE
Is it possible to have tests that selectively run for different environments (node/jsdom)?

### DISCUSSION THREAD

**Discussion by jerrythomas** [vitest-dev/vitest#2324] [Q&A]
I have a function which needs to be tested on different environments (`node`/`jsdom`). Some tests have the same `expect` for both `node` & `jsdom`. However, there are some tests that are mutually exclusive.

Below is a contrived example of how I have implemented this in different test files.

*a.spec.js*: test should only run on node
```js
import { describe, expect, test } from 'vitest'

describe('multi environment',() => {
  // @vitest-environment node
  test('this should only run in node',() => {
     expect(global.window).toBeUndefined()
  })
})
```

*b.spec.js*: test should only run on jsdom
```js
import { describe, expect, test } from 'vitest'

describe('multi environment',() => {
  // @vitest-environment jsdom
  test('this should only run in jsdom',() => {
     expect(global.window).toBeDefined()
  })
})
```

*c.spec.js*: this should run on both node & jsdom. I think this currently runs only in node not both node & jsdom
```js
import { describe, expect, test } from 'vitest'

describe('multi environment',() => {
  test('this should run on both node&jsdom', ()=>{
    expect(1+2).toEqual(3)
  })
})
```

*env.spec.js*: Mix of both node/jsdom tests
```js
import { describe, expect, test } from 'vitest'

describe('multi environment',() => {
  // @vitest-environment node
  test('this should only run in node',() => {
     expect(global.window).toBeUndefined()
  })

  // @vitest-environment jsdom
  test('this should only run in jsdom',() => {
     expect(global.window).toBeDefined()
  })

  test('this should run on both node&jsdom', ()=>{
    expect(1+2).toEqual(3)
  })
})
```

My `vitest.config.js`
```js
import { defineConfig } from 'vite'

export default defineConfig({
  test: {
    globals: true,
    environment: 'jsdom',
    coverage: {
      reporter: ['text', 'lcov'],
      all: false,
      include: ['src'],
    },
  },
})
```

Result of the run
```text
 ✓ src/a.spec.js (1)
 ❯ src/env.spec.js (3)
   ❯ multi environment (3)
     ✓ this should only run in node
     × this should only run in jsdom
     ✓ this should run on both node&jsdom
 ✓ src/b.spec.js (1)
 ✓ src/c.spec.js (1)
```

**Comment by sheremet-va** [comment]
No, it's not possible, and we do not plan to support mixed environments in one file, this doesn't fit into Vitest architecture, where _files_ are isolated.

**Reply by jerrythomas** [reply]
Ok, Thanks. Is it possible to execute the same test in multiple environments? For example `c.spec.js` being run twice once under `node` and once under `jsdom`?

**Reply by sheremet-va** [reply]
Currently no, it's not possible with Vitest configuration. You would need to run it twice. Might be enabled with this feature: https://github.com/vitest-dev/vitest/issues/256
